charger/smaevcharger.go

```diff
@@ -0,0 +1,359 @@
+func NewSmaevcharger(host string, user string, password string) (api.Charger, error) {
+	log := util.NewLogger("smaevcharger").Redact(user, password)
+
+	wb := &Smaevcharger{
+		Helper:   request.NewHelper(log),
+		log:      log,
+		host:     "http://" + host + "/api/v1",
+		user:     user,
+		password: password,
+	}
+
+	ts, err := smaevcharger.TokenSource(log, wb.host, wb.user, wb.password)
+	if err != nil {
+		return wb, err
+	}
+
+	// replace client transport with authenticated transport
+	wb.Client.Transport = &oauth2.Transport{
+		Source: ts,
+		Base:   wb.Client.Transport,
+	}
+
+	SoftwareVersion := fmt.Sprint(wb.GetParameter("Parameter.Nameplate.PkgRev"))
+	SoftwareVersionParts := strings.Split(SoftwareVersion, ".")
+	errortext := "Failed to read Charger Softwareversion"
+
+	if len(SoftwareVersionParts) < 3 {
+		return wb, errors.New(errortext)
+	}
+	tempvar1, err := strconv.Atoi(SoftwareVersionParts[0])
+	if err != nil {
+		return wb, errors.New(errortext)
+	}
+	tempvar2, err := strconv.Atoi(SoftwareVersionParts[1])
+	if err != nil {
+		return wb, errors.New(errortext)
+	}
+	tempvar3, err := strconv.Atoi(SoftwareVersionParts[2])
+	if err != nil {
+		return wb, errors.New(errortext)
+	}
+
+	if tempvar1 < 1 || tempvar2 < 2 || tempvar3 < 23 {
+		return wb, errors.New("Charger Softwareversion not supported - please update > 1.2.23R")
+	}
+
+	return wb, nil
+}
+
+// Status implements the api.Charger interface
+func (wb *Smaevcharger) Status() (api.ChargeStatus, error) {
+	StateChargerCharging := wb.GetMeasurement("Measurement.Operation.EVeh.ChaStt")
+
+	switch StateChargerCharging {
+	case smaevcharger.ConstNConNCarNChar: // No Car connectec and no charging
+		return api.StatusA, nil
+	case smaevcharger.ConstYConYCarNChar: // Car connected and no charging
+		return api.StatusB, nil
+	case smaevcharger.ConstYConYCarYChar: // Car connected and charging
+		return api.StatusC, nil
+	}
+	return api.StatusNone, fmt.Errorf("SMA EV Charger state: %s", StateChargerCharging)
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Smaevcharger) Enabled() (bool, error) {
+	StateChargerMode := wb.GetParameter("Parameter.Chrg.ActChaMod")
+
+	switch StateChargerMode {
+	case smaevcharger.ConstFastCharge: // Schnellladen - 4718
+		return true, nil
+	case smaevcharger.ConstOptiCharge: // Optimiertes Laden - 4719
+		return true, nil
+	case smaevcharger.ConstPlanCharge: // Laden mit Vorgabe - 4720
+		return true, nil
+	case smaevcharger.ConstStopCharge: // Ladestopp - 4721
+		return false, nil
+	}
+	return false, fmt.Errorf("SMA EV Charger  charge mode: %s", StateChargerMode)
+}
+
+// Enable implements the api.Charger interface
+func (wb *Smaevcharger) Enable(enable bool) error {
+	StateChargerSwitch := wb.GetMeasurement("Measurement.Chrg.ModSw")
+	if enable {
+		switch StateChargerSwitch {
+		case smaevcharger.ConstSwitchOeko: // Switch PV Loading
+			wb.SendParameter("Parameter.Chrg.ActChaMod", smaevcharger.ConstOptiCharge)
```

Comment:
Currently no error message is returned if the charger is not set to fast charge.

Reply:
error message was added